### PR TITLE
Leap year Checker

### LINKED ISSUE #52: leap_year_checker

### PR DESCRIPTION
**Pull request by sky0-0glitch**
---
#52 
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/leap_year_checker.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+"""
```
Comment (Azizsin7) [inline solutions/leap_year_checker.py:3]: add
- shebang line: #!/usr/bin/env python3
- encoding declaration: # -- coding: utf-8 --

File: solutions/leap_year_checker.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    Examples:
+        >>> is_leap_year(2000)
+        True
+        >>> is_leap_year(1900)
+        False
+        >>> is_leap_year(2024)
+        True
+        >>> is_leap_year(2023)
+        False
```
Comment (Azizsin7) [inline solutions/leap_year_checker.py:36]: there is no defensive doctest

File: solutions/tests/test_leap_year_checker.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        self.assertFalse(is_leap_year(1900))
+
+    def test_century_leap_year(self):
+        """Test a century year that is a leap year."""
+        self.assertTrue(is_leap_year(2000))
+
+    def test_non_leap_year(self):
+        """Test a common year that is not a leap year."""
+        self.assertFalse(is_leap_year(2023))
+
```
Comment (Azizsin7) [inline solutions/tests/test_leap_year_checker.py:35]: add a boundary case like large number (4000000)

File: solutions/tests/test_leap_year_checker.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+"""
```
Comment (Azizsin7) [inline solutions/tests/test_leap_year_checker.py:3]: add

shebang line: #!/usr/bin/env python3
encoding declaration: # -- coding: utf-8 --

**Review by Azizsin7 — CHANGES_REQUESTED** [review]
Function works perfect
**just add in your function:**
> shebang line: #!/usr/bin/env python3
> encoding declaration: # -- coding: utf-8 --
> defensive doctest

**For your test file:**

> shebang line: #!/usr/bin/env python3
> encoding declaration: # -- coding: utf-8 --
> boundary case like large number (4000000)

File: solutions/leap_year_checker.py (hunk 1 above)
Comment (sky0-0glitch) [reply solutions/leap_year_checker.py:3]: Added

File: solutions/leap_year_checker.py (hunk 2 above)
Comment (sky0-0glitch) [reply solutions/leap_year_checker.py:36]: Add them

File: solutions/tests/test_leap_year_checker.py (hunk 4 above)
Comment (sky0-0glitch) [reply solutions/tests/test_leap_year_checker.py:3]: Added

File: solutions/tests/test_leap_year_checker.py (hunk 3 above)
Comment (sky0-0glitch) [reply solutions/tests/test_leap_year_checker.py:35]: Added

**Review by Azizsin7 — APPROVED** [review]
perfect
